### PR TITLE
ZEPPELIN-365: Add credentials for dependency loader

### PR DESCRIPTION
**Pull request by nuttybrewer**
Allow to authenticate to Artifactory or any other private maven repository using user/pass via the dependency loader

### CODE REVIEW

**Comment by nuttybrewer** [comment]
Seems the CI environment is crashing with Bower... not related to my changes.. Should I resubmit the pull request?

**Comment by Leemoonsoo** [comment]
@nuttybrewer we're addressing https://issues.apache.org/jira/browse/ZEPPELIN-375 to make build pass on CI. Please trigger CI (close and open PR) when ZEPPELIN-375 is resolved.

File: spark/src/main/java/org/apache/zeppelin/spark/dep/DependencyContext.java [hunk 1]
```diff
@@ -74,6 +78,11 @@ public Repository addRepo(String name) {
     return rep;
   }
 
+  public void addCredential(String reponame, String user, String password) {
+    auths.put(reponame, new Authentication(user, password));
+    return;
```
Comment (felixcheung) [inline spark/src/main/java/org/apache/zeppelin/spark/dep/DependencyContext.java]: nit: no need for the return, for consistency on coding style

File: spark/src/main/java/org/apache/zeppelin/spark/dep/DependencyContext.java [hunk 2]
```diff
@@ -74,6 +78,10 @@ public Repository addRepo(String name) {
     return rep;
   }
 
+  public void addCredential(String reponame, String user, String password) {
+    auths.put(reponame, new Authentication(user, password));
```
Comment (felixcheung) [inline spark/src/main/java/org/apache/zeppelin/spark/dep/DependencyContext.java]: I think you need to clear this in reset()

**Comment by Leemoonsoo** [comment]
How about API like

```
z.addRepo("RepoName").url("RepoURL").credential("username", "password")
```

Instead of 

```
z.addRepo("RepoName").url("RepoURL")
z.addCredential("RepoName","username","password")
```

? 

I think method chaining is more consistent design here while [Repository](https://github.com/apache/incubator-zeppelin/blob/master/spark/src/main/java/org/apache/zeppelin/spark/dep/Repository.java) already has url(), snapshot() with method chaining.

**Comment by Leemoonsoo** [comment]
Looks good to me!

File: spark/src/main/java/org/apache/zeppelin/spark/dep/Repository.java [hunk 3]
```diff
@@ -51,4 +53,22 @@ public String getName() {
...
+  
+  public Repository password(String password) {
+    this.password = password;
+    return this;
+  }
+  
+  protected Authentication getAuthentication() {
+    Authentication auth = null;
+    if (this.username != null && this.password != null) {
+      auth = new Authentication(this.username, this.password);
```
Comment (felixcheung) [inline spark/src/main/java/org/apache/zeppelin/spark/dep/Repository.java]: It's possible password could be null or "" in some enterprise/internal authentication scheme perhaps? shouldn't we support `if this.username != null`

Comment (nuttybrewer) [inline spark/src/main/java/org/apache/zeppelin/spark/dep/Repository.java]: I made this change and then tested locally, Aether throws NPE when password is null. Blank password works as you described though...

Left as-is for the null check, added convenience credentials() function as well

**Comment by Leemoonsoo** [comment]
Checked update of null value handling. LGTM.
Merge if there're no more discussions.

**Pull request closed without merging**